File: Metropolis Algorithm/metropolis.m
function mcmc = metropolis(dfunc, chain, theta_init, jumpdist, space, burnin, seed)
% function to apply Metropolis Algorithm and returns a Markov Chain vector
% Input :
%     dfunc : function_handle
%       self-defined density function which take only one parameter (i.e. r.v. X)
%     chain : double, optional
%       the length of Markov Chain to be generated (default 5000)
%     theta_init : double, optional
%       initial value for the chain of θ (default 0.5)
%     jumpdist : use 'makedist' function to specify the jump distribution, optional
%       the distribution that proposed jump (Δθ) follows (default makedist('Normal','mu',0,'sigma',0.1))
%     space : 2D vector ([min, max]), optional
%       the accepted span of the parameter θ (default [-Inf,Inf])
%     burnin : double, optional
%       the length of the chain to be burned from the beginning (default 0)
%     seed : optional
%       random seed (default 'shuffle')
% Output :
%     mcmc : vector
%       generated markov chain vector [burnin + 1 : end]
arguments
    dfunc function_handle
    chain double = 5000;
    theta_init double = 0.5;
    jumpdist = makedist('Normal','mu',0,'sigma',0.1);
    space = [-Inf, Inf];
    burnin double = 0;
    seed = 'shuffle';   
end
if ~isa(dfunc,'function_handle')
    error('dfunc must be a function handle')
end

theta_cur = theta_init;
theta_freq = zeros(chain,1);
theta_freq(1) = theta_init;

rng(seed)

for i = 2:chain
    Delta_theta = random(jumpdist);
    theta_pro = theta_cur + Delta_theta;

    if theta_pro < space(1) || theta_pro > space(2)
        pmoving = 0;
    elseif dfunc(theta_pro) == 0
        pmoving = 1;
    else
        pmoving = min([1,dfunc(theta_pro)/dfunc(theta_cur)]);
    end

    if rand(1) <= pmoving
        theta_freq(i) = theta_pro;
        theta_cur = theta_pro;
    else
        theta_freq(i) = theta_cur;
    end
end

mcmc = theta_freq(burnin + 1 : end);
end

File: Metropolis Algorithm/metropolis.r
# Metropolis algorithm
metropolis <- function(chain, theta_cur, sigma = 0.2, dfunc, space_min = -Inf, space_max = Inf, burnin = 0) {
  if (!is.function(dfunc)) stop("argument dfunc is not a function")
  theta_freq <- theta_cur # theta_freq <- NULL

  while (length(theta_freq) < chain) {
    delta_theta <- rnorm(1, mean = 0, sd = sigma)
    theta_pro <- theta_cur + delta_theta
    # theta_pro <- rnorm(1, mean = theta_cur, sd = sigma)

    if (theta_pro < space_min) {
      pmoving <- 0
      theta_pro <- space_min
    } else if (theta_pro > space_max) {
      pmoving <- 0
      theta_pro <- space_max
    } else if (dfunc(theta_cur) == 0) {
      pmoving <- 1
    } else {
      pmoving <- min(dfunc(theta_pro) / dfunc(theta_cur), 1)
    }

    if (runif(1) <= pmoving) {
      theta_freq <- append(theta_freq, theta_pro)
      theta_cur <- theta_pro
    } else {
      theta_freq <- append(theta_freq, theta_cur)
    }

  }

  return(tail(theta_freq, chain - burnin))
}

# Example
# posterior function 
# prior Beta(a = 1, b = 1)
# likelihood Bernoulli: N = 20, z = 14
posterior <- function(theta, a, b, N, z) {
  dbeta(theta, z + a, N - z + b)
}
a = 1
b = 1
N = 20 
z = 14
posterior_func <- function(theta) posterior(theta, a, b, N, z)

dposterior1 <- metropolis(chain = 5000, theta_cur = 0.1, dfunc = posterior_func, space_min = 0, space_max = 1, burnin = 2500)
dposterior2 <- metropolis(chain = 5000, theta_cur = 0.9, dfunc = posterior_func, space_min = 0, space_max = 1, burnin = 2500)
dposterior3 <- metropolis(chain = 5000, theta_cur = 0.5, dfunc = posterior_func, space_min = 0, space_max = 1, burnin = 2500)